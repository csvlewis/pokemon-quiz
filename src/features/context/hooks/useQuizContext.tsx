import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
