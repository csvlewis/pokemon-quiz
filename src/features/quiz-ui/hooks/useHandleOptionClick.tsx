import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
