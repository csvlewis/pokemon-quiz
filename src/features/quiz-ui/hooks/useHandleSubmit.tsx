import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
