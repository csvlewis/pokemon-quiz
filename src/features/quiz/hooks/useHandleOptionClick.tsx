import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
