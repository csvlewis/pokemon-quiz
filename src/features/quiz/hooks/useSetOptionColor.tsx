import { OverridableStringUnion } from "@mui/types";

import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
