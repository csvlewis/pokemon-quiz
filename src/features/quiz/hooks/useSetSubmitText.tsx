import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetSubmitText(): () => string {
  const { currentQuestion, displayAnswer, questions } = useQuizContext();

  return () => {
    let text: string = "Submit";

    if (displayAnswer) {
      if (currentQuestion >= questions.length) {
        text = "See Results";
      } else text = "Next Question";
    }

    return text;
  };
}
