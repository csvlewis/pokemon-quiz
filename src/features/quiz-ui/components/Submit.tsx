import { Button, Typography } from "@mui/material";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useHandleSubmit } from "@/features/quiz-ui/hooks/useHandleSubmit";
import { useSetSubmitText } from "../hooks/useSetSubmitText";

export function Submit(): React.ReactNode {
  const { quizInProgress, selectedAnswer } = useQuizContext();
  const handleSubmit: () => void = useHandleSubmit();
  const setSubmitText: () => string = useSetSubmitText();

  if (quizInProgress)
    return (
      <Button
        disabled={selectedAnswer.text === undefined}
        variant="contained"
        onClick={() => handleSubmit()}
      >
        <Typography>{setSubmitText()}</Typography>
      </Button>
    );
}
