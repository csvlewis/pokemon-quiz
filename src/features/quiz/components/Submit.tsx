import { Button, Typography } from "@mui/material";

import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useHandleSubmit } from "@/features/quiz/hooks/useHandleSubmit";
import { useSetSubmitText } from "@/features/quiz/hooks/useSetSubmitText";

export function Submit(): React.ReactNode {
  const { quizInProgress, selectedAnswer } = useQuizContext();
  const handleSubmit: () => void = useHandleSubmit();
  const setSubmitText: () => string = useSetSubmitText();

  if (quizInProgress)
    return (
      <Button
        disabled={selectedAnswer.text === undefined}
        variant="contained"
        onClick={() => handleSubmit()}
      >
        <Typography>{setSubmitText()}</Typography>
      </Button>
    );
}
