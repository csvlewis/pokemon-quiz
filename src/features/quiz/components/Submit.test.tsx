import React from "react";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Submit } from "./Submit";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("Submit", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      quizInProgress: true,
      selectedAnswer: { text: "", isCorrect: false },
    });

    const { container } = render(<Submit />);
    expect(container).toMatchSnapshot();
  });
});
