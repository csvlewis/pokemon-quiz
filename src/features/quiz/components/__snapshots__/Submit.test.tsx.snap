// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submit should render correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      Submit
    </p>
  </button>
</div>
`;
